core/src/main/java/org/apache/accumulo/core/security/crypto/CryptoServiceFactory.java

```diff
@@ -16,30 +16,30 @@
  */
 package org.apache.accumulo.core.security.crypto;
 
+import java.util.concurrent.atomic.AtomicReference;
+
 import org.apache.accumulo.core.conf.AccumuloConfiguration;
 import org.apache.accumulo.core.conf.Property;
 import org.apache.accumulo.core.spi.crypto.CryptoService;
 import org.apache.accumulo.core.spi.crypto.CryptoService.CryptoException;
 import org.apache.accumulo.start.classloader.vfs.AccumuloVFSClassLoader;
 
 public class CryptoServiceFactory {
-  private static CryptoService singleton = null;
+  private static AtomicReference<CryptoService> singleton = new AtomicReference<>();
 
   /**
    * Load the singleton class configured in {@link Property#INSTANCE_CRYPTO_SERVICE}
    */
   public static CryptoService getConfigured(AccumuloConfiguration conf) {
+    CryptoService cyptoService = singleton.get();
     String configuredClass = conf.get(Property.INSTANCE_CRYPTO_SERVICE.getKey());
-    if (singleton == null) {
-      singleton = loadCryptoService(configuredClass);
-      singleton.init(conf.getAllPropertiesWithPrefix(Property.INSTANCE_CRYPTO_PREFIX));
-    } else {
-      if (!singleton.getClass().getName().equals(configuredClass)) {
-        singleton = loadCryptoService(configuredClass);
-        singleton.init(conf.getAllPropertiesWithPrefix(Property.INSTANCE_CRYPTO_PREFIX));
-      }
+    if (cyptoService == null || !cyptoService.getClass().getName().equals(configuredClass)) {
+      CryptoService newCryptoService = loadCryptoService(configuredClass);
+      newCryptoService.init(conf.getAllPropertiesWithPrefix(Property.INSTANCE_CRYPTO_PREFIX));
+      singleton.compareAndSet(cyptoService, newCryptoService);
+      return singleton.get();
```

Comment:
May be better to return newCryptoService here, because in the case that compare and set fails the crypto service from singleton.get() may not match config.  For example if there are different configs in the JVM (for example in the case of someone using the RFile API with multiple configs).

Reply:
Hmm OK I hadn't thought of that.

Reply:
Wondering if I should check the compareAndSet result.  And if so what should I do if it fails?  Is it worth logging a warning?

Reply:
> Wondering if I should check the compareAndSet result. And if so what should I do if it fails? Is it worth logging a warning?

I don't think its worth logging or we care if it fails, its kinda of a cache w/o any strong guarantees.  Don't even have to do compare and set, could just set.   W/o sync there is no way to prevent multiple service objects for the same config.  Even w/ sync, if there are multiple active configs in the JVM there is no way to prevent multiple services objects for the same config w/o sync+map (where map key is something unique derived from config).

Reply:
For a single config in the JVM, could avoid multiple service object for same config with compareAndSet + loop.  This would not help with multiple config situation though.

Reply:
This still doesn't prevent concurrency issues though does it?  There is still the possibility of a race condition where CryptoService.init() gets called twice.